src/utils/include/utils/LCA.hpp

```diff
@@ -15,100 +12,99 @@ class LCA
 
     std::unordered_map< std::string, std::vector< std::string > > m_parents;
     std::vector< int >                                            m_euler;
     std::vector< int >                                            m_depth;
-    std::unique_ptr< int[] >                                      m_firstAppearance;
+    std::vector< int >                                            m_firstAppearance;
     int                                                           m_vertices = 0;
     std::unordered_map< std::string, int >                        m_encode;
     std::unordered_map< int, std::string >                        m_decode;
-    std::unique_ptr< std::unique_ptr< int[] >[] >                 m_M;
+    std::vector< std::vector< int > >                             m_M;
 };
 
-inline void LCA::addEdge( std::string father, std::string son )
+inline void LCA::addEdge( const std::string& father, const std::string& son )
 {
     if ( m_encode.count( father ) == 0 )
     {
         m_encode.insert( { father, m_vertices } );
         m_decode.insert( { m_vertices, father } );
-        m_vertices++;
+        ++m_vertices;
     }
+
     if ( m_encode.count( son ) == 0 )
     {
         m_encode.insert( { son, m_vertices } );
         m_decode.insert( { m_vertices, son } );
-        m_vertices++;
+        ++m_vertices;
     }
+
     if ( m_parents.count( father ) == 0 )
     {
-        std::vector< std::string > children;
-        children.push_back( son );
-        m_parents[father] = children;
+        m_parents[father] = { son };
     }
     else
     {
-        m_parents[father].push_back( son );
+        m_parents[father].emplace_back( son );
     }
 }
 
-inline void LCA::depthFirstSearch( std::string current, int depth )
+inline void LCA::depthFirstSearch( const std::string& current, int depth )
 {
+    const auto currentEncoded = m_encode[current];
+
     // marking first appearance for current node
-    if ( m_firstAppearance[m_encode[current]] == -1 )
+    if ( m_firstAppearance[currentEncoded] == -1 )
     {
-        m_firstAppearance[m_encode[current]] = m_euler.size();
+        m_firstAppearance[currentEncoded] = m_euler.size();
     }
+
     // pushing root to euler walk
-    m_euler.push_back( m_encode[current] );
+    m_euler.push_back( currentEncoded );
     // pushing depth of current node
     m_depth.push_back( depth );
-    for ( unsigned int i = 0; i < m_parents[current].size(); i++ )
+
+    for ( const auto& node : m_parents[current] )
     {
-        depthFirstSearch( m_parents[current][i], depth + 1 );
-        m_euler.push_back( m_encode[current] );
+        depthFirstSearch( node, depth + 1 );
+        m_euler.push_back( currentEncoded );
         m_depth.push_back( depth );
     }
 }
 
 inline void LCA::doEulerWalk()
 {
-    m_firstAppearance = std::make_unique< int[] >( m_vertices );
-    for ( int i = 0; i < m_vertices; i++ )
-    {
-        m_firstAppearance[i] = -1;
-    }
+    m_firstAppearance.resize( m_vertices, -1 );
```

Comment:
For the sake of being clear about intentions and to avoid magic numbers, I would declare a private `constexpr int invalid_first_appearance = -1`. This then would be more like:

```
m_firstAppearance.resize( m_vertices, invalid_first_appearance );
```

And then, the validation test in `LCA::depthFirstSearch( const std::string&, int )` would look like:

```
if ( m_firstAppearance[currentEncoded] == invalid_first_appearance )
{
    m_firstAppearance[currentEncoded] = m_euler.size();
}
```